Lib/tarfile.py

```diff
@@ -538,6 +538,7 @@ def _read(self, size):
             return self.__read(size)
 
         c = len(self.dbuf)
+        t = [self.dbuf]
         while c < size:
             buf = self.__read(self.bufsize)
             if not buf:
```

Comment:
How about bypassing `self.__read()`?

```python
        while c < size:
            if self.buf:
                buf = self.buf
                self.buf = b""
            else:
                buf = self.fileobj.read(self.bufsize)
                if not buf:
                    break
```

Reply:
For compressed streams your suggestion works.  However, there is one case where `self.__read()` is called with large `size`: When you open the stream uncompressed with `"r|"`. For this special case, you still need the optimization in `self.__read()`.

https://github.com/python/cpython/blob/3c452404ae178b742967589a0bb4a5ec768d76e0/Lib/tarfile.py#L537-L538

Reply:
Regardless optimization `self.__read()`, double (and unaligned) buffering is totally useless.

Reply:
Yes, I agree. It is a bit twisted to handle the different cases of decompressing versus direct read.  So far, this is just a minimal fix of the performance regression.

Reply:
You're right.
To backport this, fix only repeated bytes += bytes and keep other behavior as-is.